### PR TITLE
Refactor MeshRenderer and Renderer

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Other information:

### CODE REVIEW

File: packages/core/src/Renderer.ts [hunk 1]
```diff
@@ -79,6 +97,146 @@ export abstract class Renderer extends Component {
...
+  /**
+   * Get the first material by index.
+   * @param index - Material index
+   * @returns Material
+   */
+  getMaterial(index: number): Material | null;
+
+  getMaterial(index: number = 0): Material | null {
+    const materials = this._materials;
+    return materials.length > 0 ? materials[index] : null;
```
Comment (zhuxudong) [inline packages/core/src/Renderer.ts:144]: `materials[index]` may be `undefined`

File: packages/core/src/Renderer.ts [hunk 2]
```diff
@@ -79,6 +97,146 @@ export abstract class Renderer extends Component {
...
+    materialsInstanced.length !== 0 && (materialsInstanced.length = 0);
+
+    for (let i = 0; i < count; i++) {
+      const internalMaterial = internalMaterials[i];
+      const material = materials[i];
+      if (internalMaterial !== material) {
+        internalMaterials[i] = material;
+        internalMaterial && internalMaterial._addRefCount(-1);
+        material && material._addRefCount(1);
+      }
```
Comment (zhuxudong) [inline packages/core/src/Renderer.ts:232]: Can this be replaced with `this.setMaterial(i,material)`?

File: packages/core/src/Renderer.ts (hunk 1 above)
Comment (singlecoder) [reply packages/core/src/Renderer.ts:144]: can use materials.length > index

File: packages/core/src/Renderer.ts [hunk 3]
```diff
@@ -79,6 +97,146 @@ export abstract class Renderer extends Component {
...
+  setMaterial(material: Material);
+
+  /**
+   * Set material by index.
+   * @param index - Material index
+   * @param material - The material
+   */
+  setMaterial(index: number, material: Material);
+
+  setMaterial(indexOrMaterial: number | Material, material: Material = null) {
```
Comment (singlecoder) [inline packages/core/src/Renderer.ts:160]: Can use setMaterial(material: Material, indexOfMaterial: number = 0)

File: packages/core/src/particle/ParticleRenderer.ts [hunk 4]
```diff
@@ -590,8 +588,8 @@ export class ParticleRenderer extends GeometryRenderer {
     return material;
   }
 
-  private _createGeometry(): BufferGeometry {
-    const geometry = new BufferGeometry(this._entity.engine, "particleGeometry");
+  private _createGeometry(): Mesh {
```
Comment (singlecoder) [inline packages/core/src/particle/ParticleRenderer.ts:591]: Rename _createGeometry to _createMesh should be better?

File: packages/core/src/Renderer.ts (hunk 3 above)
Comment (GuoLei1990) [reply packages/core/src/Renderer.ts]: Not Good! If no function overload, I will design like that. If use function overload,this is better for developers.

File: packages/core/src/Renderer.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/core/src/Renderer.ts:144]: You are right! it's a bug!

File: packages/core/src/Renderer.ts (hunk 2 above)
Comment (GuoLei1990) [reply packages/core/src/Renderer.ts:231]: Can! But some customized optimization must be lost.

File: packages/core/src/particle/ParticleRenderer.ts (hunk 4 above)
Comment (GuoLei1990) [reply packages/core/src/particle/ParticleRenderer.ts:591]: Yes！

File: packages/core/src/Renderer.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/core/src/Renderer.ts]: fixed!

File: packages/core/src/animation/Animation.ts [hunk 5]
```diff
@@ -372,7 +372,6 @@ export class Animation extends Component {
 
       if (path === "weights") {
         // SkinnedMeshRenderer.
-        (targetObject as SkinnedMeshRenderer).setWeights(val as any);
```
Comment (singlecoder) [inline packages/core/src/animation/Animation.ts:375]: 这个去掉不会影响现有逻辑么

Comment (GuoLei1990) [reply packages/core/src/animation/Animation.ts:375]: No affect,before is use for morph

**Pull request merged**